### PR TITLE
dnsdist: Implement Lua custom web endpoints

### LINKED ISSUE #9120: dnsdist: add support for a Lua HTTP API handler
- Program: dnsdist
- Issue type: Feature request

### Short description

It would be nice if dnsdist supported adding an API HTTP handler in Lua.

### Usecase

This would allow HTTP clients to update the Lua configuration in arbitrary - but controlled - ways without having to implement custom API endpoints for every possible feature. Currently this can only be achieved by opening a socket to the dnsdist console and writing Lua function calls to it.

### Description

For example, dnsdist could define a `/api/v1/servers/localhost/custom[/....]` endpoint that is handled in Lua. When an HTTP request is received, dnsdist calls a user-defined `customhttphandler` Lua function to handle the request.

The Lua callback function could look as follows:

```lua
function customhttphandler(req)
    print(req.method, req.path, req.body)
    return 200, {["Content-Type"] = "text/plain"}, "hello world!"
end 
```

The request object would be a table for easy future extensibility without an explosion of positional arguments.  Suggested attributes to include in the request object:

- method: the HTTP method
- path: the full request path
- body: the raw request body as string
- params: parsed GET params
- form: parsed POST values if this is a POST request and the content type is `application/x-www-form-urlencoded`
- subpath: the subpath under `/api/v1/servers/localhost/custom`, allowing for custom routing in Lua.

The response args are:

- The HTTP status code
- An optional table of extra HTTP headers to set on the response (nil allowed)
- A string with the response contents

In order to limit the time spent in the Lua code (which will block dnsdist), dnsdist buffers the whole request with body before calling the Lua code. This processing is all done in a separate thread. For the same reason, the Lua handler returns the whole response at once, instead of streaming the response as is common in modern frameworks.

If the `customhttphandler` callback is not defined, dnsdist returns a 404 and an appropriate message.

The HTTP request is subject to API key validation.

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Closes #9120.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdistdist/docs/reference/web.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  .. attribute:: WebRequest.version
+
+    The HTTP version of this query, as an integer.
+
+.. class:: WebResponse
+
+  Represent a HTTP response.
+
+  .. attribute:: WebRequest.body
```
Comment (Habbie) [inline pdns/dnsdistdist/docs/reference/web.rst:42]:
```suggestion
  .. attribute:: WebResponse.body
```

? (and two more)

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 2]
```diff
@@ -349,6 +349,22 @@ Webserver configuration
   * ``custom_headers={[str]=str,...}``: map of string - Allows setting custom headers and removing the defaults.
   * ``acl=newACL``: string - List of IP addresses, as a string, that are allowed to open a connection to the web server. Defaults to "127.0.0.1, ::1".
 
+.. function:: registerWebHandler(path, handler)
+
+  .. versionadded: 1.6.0
+
+  Register a function named ``handler`` that will be called for every query sent to the exact ``path`` path. The function will receive a :class:`WebRequest` object
+  and a :class:`WebResponse` object, representing respectively the HTTP request received and the HTTP response to send.
```
Comment (Habbie) [inline pdns/dnsdistdist/docs/reference/config.rst:357]: could use an example of handling a simple request and setting resp.body etc.

Comment (rgacogne) [reply pdns/dnsdistdist/docs/reference/config.rst:357]: That makes total sense, will do!

**Comment by rgacogne** [comment]
Weird test-auth-regress-bind error in bind-dnssec-nsec3-both, during verify-dnssec-zone:
```
Nov 06 10:40:12 AXFR of domain 'tsig.com' allowed: client IP 127.0.0.1:46172 is in allow-axfr-ips
AXFR failed.
Syntax error, could not parse the RR's TTL at 0
Nov 06 10:40:12 Received serial number updates for 1 zone, had 0 timeouts
```

**Comment by wojas** [comment]
Nice!

**Pull request merged**